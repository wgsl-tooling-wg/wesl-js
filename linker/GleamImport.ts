import {
  disablePreParse,
  kind,
  makeEolf,
  matchOneOf,
  NoTags,
  opt,
  or,
  Parser,
  repeatPlus,
  seq,
  setTraceName,
  TagRecord,
  tokenMatcher,
  tokens,
  tokenSkipSet,
  tracing,
  withSepPlus,
  withTags,
} from "@wesl/mini-parse";
import { TreeImportElem } from "./AbstractElems.ts";
import {
  ImportTree,
  PathSegment,
  SegmentList,
  SimpleSegment,
  Wildcard,
} from "./ImportTree.ts";
import { digits, eol, word } from "./MatchWgslD.ts";
import { makeElem } from "./ParseSupport.ts";

const gleamImportSymbolSet = "/ { } , ( ) .. . * ;";
const gleamImportSymbol = matchOneOf(gleamImportSymbolSet);

const skipWsSet = new Set(["ws"]);
function skipWs<V, T extends TagRecord>(p: Parser<V, T>): Parser<V, T> {
  return tokenSkipSet(skipWsSet, p);
}
function noSkipWs<V, T extends TagRecord>(p: Parser<V, T>): Parser<V, T> {
  return tokenSkipSet(null, p);
}
const ws = /\s+/;

export const gleamImportTokens = tokenMatcher({
  ws,
  gleamImportSymbol,
  word,
  digits,
});

export const packageTokens = tokenMatcher({
  ws,
  pkg: /[a-zA-Z_][\w-]*/, // LATER follow spec
  other: /.+/,
});

export const eolTokens = tokenMatcher({
  ws: /[ \t]+/, // don't include \n, for eolf
  eol,
});

const eolf = disablePreParse(
  makeEolf(eolTokens, gleamImportTokens.ws).traceName("gleam_eolf"),
);
const wordToken = kind(gleamImportTokens.word);
const pkgToken = kind(packageTokens.pkg);

// forward references (for mutual recursion)
let pathTail: Parser<PathSegment[], NoTags> = null as any;
let packagePath: Parser<PathSegment[], NoTags> = null as any;

const simpleSegment = wordToken.map(r => {
  return new SimpleSegment(r.value);
});

const itemImport = withTags(
  seq(
    wordToken.tag("segment"),
    skipWs(opt(seq("as", wordToken.tag("as")))),
  ).map(r => {
    const segment = r.tags.segment[0];
    return new SimpleSegment(segment, r.tags.as?.[0]);
  }),
);

const starImport = seq(
  skipWs(seq("*", opt(seq("as", wordToken.tag("as"))))),
).map(r => new Wildcard(r.tags.as?.[0]));

const collectionItem = or(
  () => packagePath,
  itemImport.map(r => [r.value]),
);

const importCollection = withTags(
  seq(
    "{",
    skipWs(
      seq(
        withSepPlus(",", () => collectionItem).tag("list"),
        "}", //
      ),
    ),
  ).map(r => {
    const elems = r.tags.list.flat().map(l => new ImportTree(l));
    return new SegmentList(elems);
  }),
);

const pathSegment = or(simpleSegment, importCollection);

const pathExtends = withTags(
  seq(simpleSegment.tag("s"), "/", () => pathTail.tag("s")).map(r =>
    r.tags.s.flat(),
  ),
);

/** The tail covers the part of the import path after the prefix */
pathTail = withTags(
  or(
    pathExtends,
    or(importCollection, itemImport, starImport).map(r => [r.value]),
  ).map(r => {
    return r.value.flat();
  }),
);

// The prefix covers the import path until the point we could import an item
// so ../foo or foo/

const relativeSegment = withTags(
  seq(or(".", "..").tag("dir"), "/").map(r => new SimpleSegment(r.tags.dir[0])),
);

const relativePrefix = withTags(
  seq(
    repeatPlus(relativeSegment.tag("seg")),
    simpleSegment.tag("seg"),
    "/",
  ).map(r => {
    return r.tags.seg;
  }),
);

const relativePath = withTags(
  seq(relativePrefix.tag("p"), pathTail.tag("p")).map(r => r.tags.p.flat()),
);

const packagePrefix = withTags(
  seq(tokens(packageTokens, pkgToken.tag("pkg")), "/").map(r => [
    new SimpleSegment(r.tags.pkg[0]),
  ]),
);

packagePath = seq(packagePrefix, pathTail).map(r => r.value.flat());

const fullPath = noSkipWs(
  seq(kind(gleamImportTokens.ws), or(relativePath, packagePath).tag("path")),
).map(r => {
  return new ImportTree(r.tags.path.flat());
});

/** parse a Gleam style wgsl import statement. */
export const gleamImport = withTags(
  tokens(
    gleamImportTokens,
    seq("import", fullPath.tag("imports"), opt(";"), eolf).map(r => {
      const e = makeElem("treeImport", r, ["imports"]) as TreeImportElem;
      r.app.state.push(e);
    }),
  ),
);

if (tracing) {
  const names: Record<string, Parser<unknown, TagRecord>> = {
    simpleSegment,
    itemImport,
    starImport,
    importCollection,
    pathSegment,
    pathExtends,
    pathTail,
    relativeSegment,
    relativePrefix,
    relativePath,
    packagePrefix,
    packagePath,
    fullPath,
    gleamImport,
  };

  Object.entries(names).forEach(([name, parser]) => {
    setTraceName(parser, name);
  });
}
