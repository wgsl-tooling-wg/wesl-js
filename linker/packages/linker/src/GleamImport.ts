import {
  disablePreParse,
  kind,
  makeEolf,
  matchOneOf,
  NoTags,
  opt,
  or,
  Parser,
  repeatPlus,
  seq,
  setTraceName,
  TagRecord,
  tagScope,
  tokenMatcher,
  tokens,
  tokenSkipSet,
  tracing,
  withSepPlus,
  withTags,
} from "mini-parse";
import { TreeImportElem } from "./AbstractElems.js";
import {
  ImportTree,
  PathSegment,
  SegmentList,
  SimpleSegment,
  Wildcard,
} from "./ImportTree.js";
import { digits, eol, word } from "./MatchWgslD.js";
import { makeElem } from "./ParseSupport.js";
import { importElem, importList, importSegment } from "./WESLCollect.js";

const gleamImportSymbolSet = "/ { } , ( ) .. . * ; @ #"; // Had to add @ and # here to get the parsing tests to work. Weird.
const gleamImportSymbol = matchOneOf(gleamImportSymbolSet);

const skipWsSet = new Set(["ws"]);
function skipWs<V, T extends TagRecord>(p: Parser<V, T>): Parser<V, T> {
  return tokenSkipSet(skipWsSet, p);
}
function noSkipWs<V, T extends TagRecord>(p: Parser<V, T>): Parser<V, T> {
  return tokenSkipSet(null, p);
}
const ws = /\s+/;

export const gleamImportTokens = tokenMatcher({
  ws,
  gleamImportSymbol,
  word,
  digits,
});

export const packageTokens = tokenMatcher({
  ws,
  pkg: /[a-zA-Z_][\w-]*/, // LATER follow spec
  other: /.+/,
});

export const eolTokens = tokenMatcher({
  ws: /[ \t]+/, // don't include \n, for eolf
  eol,
});

const eolf = disablePreParse(
  makeEolf(eolTokens, gleamImportTokens.ws).traceName("gleam_eolf"),
);
const wordToken = kind(gleamImportTokens.word);
const pkgToken = kind(packageTokens.pkg);

// forward references for mutual recursion
let pathTail: Parser<PathSegment[], NoTags> = null as any;
let packagePath: Parser<PathSegment[], NoTags> = null as any;

const simpleSegment = tagScope(
  wordToken
    .ptag("segment")
    .collect(importSegment)
    .map(r => {
      return new SimpleSegment(r.value);
    }),
);

const itemImport = tagScope(
  withTags(
    seq(
      wordToken.tag("segment").ptag("segment"),
      skipWs(opt(seq("as", wordToken.ptag("as").tag("as")))),
    )
      .collect(importSegment)
      .map(r => {
        const segment = r.tags.segment[0];
        return new SimpleSegment(segment, r.tags.as?.[0]);
      }),
  ),
);

const starImport = seq(
  skipWs(seq("*", opt(seq("as", wordToken.tag("as"))))),
).map(r => new Wildcard(r.tags.as?.[0]));

const collectionItem = or(
  () => packagePath,
  itemImport.map(r => [r.value]),
);

const importCollection = tagScope(
  withTags(
    seq(
      "{",
      skipWs(
        seq(
          withSepPlus(",", () => collectionItem)
            .ptag("list")
            .tag("list"),
          "}", //
        ),
      ),
    )
      .collect(importList)
      .map(r => {
        const elems = r.tags.list.flat().map(l => new ImportTree(l));
        return new SegmentList(elems);
      }),
  ),
);

const pathExtends = withTags(
  seq(simpleSegment.tag("s"), "/", () => pathTail.tag("s")).map(r =>
    r.tags.s.flat(),
  ),
);

/** The tail covers the part of the import path after the prefix */
pathTail = withTags(
  or(
    pathExtends,
    or(importCollection, itemImport, starImport).map(r => [r.value]),
  ).map(r => {
    return r.value.flat();
  }),
).ctag("seg");

const relativeSegment = tagScope(
  withTags(
    seq(or(".", "..").ptag("segment").tag("dir"), "/")
      .collect(importSegment)
      .map(r => new SimpleSegment(r.tags.dir[0])),
  ),
);

// The prefix covers the import path until the last item
const relativePrefix = seq(
  repeatPlus(relativeSegment.ctag("seg").tag("seg")),
  simpleSegment.ctag("seg").tag("seg"),
  "/",
).map(r => {
  return r.tags.seg;
});

const relativePath = withTags(
  seq(relativePrefix.tag("p"), pathTail.tag("p")).map(r => r.tags.p.flat()),
);

const packagePrefix = withTags(
  seq(tokens(packageTokens, pkgToken.tag("pkg")), "/").map(r => [
    new SimpleSegment(r.tags.pkg[0]),
  ]),
);

packagePath = seq(packagePrefix, pathTail).map(r => r.value.flat());

const fullPath = noSkipWs(
  seq(kind(gleamImportTokens.ws), or(relativePath, packagePath).tag("path")),
).map(r => {
  return new ImportTree(r.tags.path.flat());
});

/** parse a Gleam style wgsl import statement. */
export const gleamImport = tagScope(
  withTags(
    tokens(
      gleamImportTokens,
      seq("import", fullPath.tag("imports"), opt(";"), eolf)
        .collect(importElem())
        .map(r => {
          const e = makeElem("treeImport", r, ["imports"]) as TreeImportElem;
          r.app.stable.elems.push(e);
        }),
    ),
  ),
);

if (tracing) {
  const names: Record<string, Parser<unknown, TagRecord>> = {
    simpleSegment,
    itemImport,
    starImport,
    importCollection,
    pathExtends,
    pathTail,
    relativeSegment,
    relativePrefix,
    relativePath,
    packagePrefix,
    packagePath,
    fullPath,
    gleamImport,
  };

  Object.entries(names).forEach(([name, parser]) => {
    setTraceName(parser, name);
  });
}
