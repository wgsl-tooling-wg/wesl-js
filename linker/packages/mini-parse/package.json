{
  "name": "mini-parse",
  "version": "0.5.0-d1",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "build": "run-s build:main build:util build:minified build:brotli",
    "build:main": "vite build",
    "build:util": "vite -c test-util.vite.config.ts build",
    "build:minified": "vite -c minified.vite.config.ts build",
    "build:brotli": "brotli dist/minified.cjs && ls -l dist/minified.cjs.br",
    "format": "prettier . --write",
    "lint": "eslint src",
    "organize": "organize-imports-cli tsconfig.json",
    "prepublishOnly": "run-s build",
    "test": "vitest",
    "test:once": "vitest run"
  },
  "type": "module",
  "repository": "github:wgsl-tooling-wg/wgsl-linker",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    },
    "./minified": {
      "types": "./dist/index.d.ts",
      "import": "./dist/minified.js",
      "require": "./dist/minified.cjs"
    },
    "./test-util": {
      "import": "./dist/testUtil.js",
      "types": "./dist/test-util/testUtil.d.ts"
    }
  },
  "devDependencies": {
    "@types/node": "^22.5.0",
    "@typescript-eslint/eslint-plugin": "^8.2.0",
    "@typescript-eslint/parser": "^8.2.0",
    "@vitest/ui": "^2.0.5",
    "berry-pretty": "^0.0.5",
    "esbuild": "^0.23.1",
    "eslint": "^9.9.0",
    "eslint-config-prettier": "^9.1.0",
    "npm-run-all": "^4.1.5",
    "organize-imports-cli": "^0.10.0",
    "prettier-eslint-cli": "^8.0.1",
    "rollup-plugin-visualizer": "^5.12.0",
    "terser": "^5.31.6",
    "typescript": "^5.5.4",
    "vite": "^5.4.2",
    "vite-plugin-dts": "^4.0.3",
    "vite-tsconfig-paths": "^5.0.1",
    "vitest": "^2.0.5"
  }
}
