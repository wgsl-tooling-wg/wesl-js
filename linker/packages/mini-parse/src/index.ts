export * from "./MatchingLexer.js";
export * from "./Parser.js";
export * from "./ParserCollect.js";
export * from "./ParserCombinator.js";
export * from "./ParserLogging.js";
export * from "./ParserToString.js";
export * from "./ParserTracing.js";
export * from "./Span.js";
export * from "./SrcMap.js";
export * from "./SrcMapBuilder.js";
export * from "./TokenMatcher.js";
export * from "./WrappedLog.js";
