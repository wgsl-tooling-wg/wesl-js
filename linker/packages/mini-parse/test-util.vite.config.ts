/// <reference types="vitest/config" />
import { LibraryOptions, defineConfig } from "vite";
import { baseViteConfig } from "./base.vite.config.js";
import { resolve } from "path";

const config = baseViteConfig();
config.build.emptyOutDir = false;
const lib = config.build.lib as LibraryOptions;
lib.name = "mini-parse-test-util";
lib.formats = ["es"];
lib.entry = [resolve(__dirname, "./src/test-util/index.ts")];
lib.fileName = "testUtil";

export default defineConfig(config);
