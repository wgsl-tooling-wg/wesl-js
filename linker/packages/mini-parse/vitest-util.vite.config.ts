/// <reference types="vitest/config" />
import { LibraryOptions, defineConfig } from "vite";
import { baseViteConfig } from "./base.vite.config.js";
import { resolve } from "path";

const config = baseViteConfig();
config.build.emptyOutDir = false;
const lib = config.build.lib as LibraryOptions;
lib.name = "mini-parse-vitest-util";
lib.formats = ["es"];
lib.entry = [resolve(__dirname, "./src/vitest-util/index.ts")];
lib.fileName = "vitestUtil";

export default defineConfig(config);