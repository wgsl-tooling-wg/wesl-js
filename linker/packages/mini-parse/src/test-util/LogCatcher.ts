import { withLogger, withLoggerAsync } from "mini-parse";

export interface LogCatcher {
  /** tests can use this to replace console.log with a log capturing function */
  log: (...params: any[]) => void;
  logged: () => string;
}

export function logCatch(): LogCatcher {
  const lines: string[] = [];
  function log(...params: any[]): void {
    lines.push(params.join(" "));
  }
  function logged(): string {
    return lines.join("\n");
  }
  return { log, logged };
}

export function withLogSpy(fn: () => void): string {
  const catcher = logCatch();
  withLogger(c => catcher.log(c), fn);
  return catcher.logged();
}

export async function withLogSpyAsync(
  fn: () => Promise<void>,
): Promise<string> {
  const catcher = logCatch();
  await withLoggerAsync(c => catcher.log(c), fn);
  return catcher.logged();
}
