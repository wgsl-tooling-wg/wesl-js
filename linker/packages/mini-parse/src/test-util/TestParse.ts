import {
  enableTracing,
  matchingLexer,
  matchOneOf,
  NoTags,
  OptParserResult,
  Parser,
  TagRecord,
  TokenMatcher,
  tokenMatcher,
  tracing,
  _withBaseLogger,
  AppState,
} from "mini-parse";
import { expect } from "vitest";
import { logCatch } from "./LogCatcher.js";

const symbolSet =
  "& && -> @ / ! [ ] { } : , = == != > >= < << <= % - -- ' \"" +
  ". + ++ | || ( ) ; * ~ ^ // /* */ += -= *= /= %= &= |= ^= >>= <<= <<";
export const testTokens = tokenMatcher({
  directive: /#[a-zA-Z_]\w*/,
  word: /[a-zA-Z_]\w*/,
  attr: /@[a-zA-Z_]\w*/,
  symbol: matchOneOf(symbolSet),
  digits: /\d+/,
  ws: /\s+/,
});

export interface TestParseResult<T, N extends TagRecord = NoTags, S = any> {
  parsed: OptParserResult<T, N>;
  position: number;
  stable: any;
}

/** utility for testing parsers */
export function testParse<T, N extends TagRecord = NoTags, S = any>(
  p: Parser<T, N>,
  src: string,
  tokenMatcher: TokenMatcher = testTokens,
  appState: AppState<S> = { context: {} as S, stable: [] },
): TestParseResult<T, N, S> {
  const lexer = matchingLexer(src, tokenMatcher);
  const parsed = p.parse({ lexer, appState: appState, maxParseCount: 1000 });
  return { parsed, position: lexer.position(), stable: appState.stable };
}

/** run a test function and expect that no error logs are produced */
export function expectNoLog<T>(fn: () => T): T {
  const { log, logged } = logCatch();
  let result: T | undefined = undefined;

  try {
    result = _withBaseLogger(log, fn);
  } finally {
    if (logged()) {
      console.log(logged());
    }
    expect(logged()).toBe("");
  }
  return result;
}

/** run a test with tracing facility disabled
 * (e.g. if the tracing facility might interfere with the test) */
export function withTracingDisabled(fn: () => void): void {
  const tracingWasEnabled = tracing;
  enableTracing(false);
  try {
    fn();
  } finally {
    enableTracing(tracingWasEnabled);
  }
}
