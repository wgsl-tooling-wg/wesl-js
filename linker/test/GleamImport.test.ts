import { testParse, TestParseResult } from "@wesl/mini-parse/test-util";
import { expect, test } from "vitest";
import { assertSnapshot } from "@std/testing/snapshot";
import { gleamImport } from "../GleamImport.ts";

function expectParses(ctx: Deno.TestContext): TestParseResult<void> {
  const result = testParse(gleamImport, ctx.name);
  expect(result.parsed).not.toBeNull();
  return result;
}
/* ------  success cases  -------   */

test("import ./foo/bar;", (ctx) => {
  const result = expectParses(ctx);
  expect(result.position).toBe(ctx.name.length); // consume semicolon (so that linking will remove it)
});

test("import foo-bar/boo", (ctx) => {
  expectParses(ctx);
});

/**  ----- extraction tests -----  */
test("import foo/bar", async (ctx) => {
  const { appState } = expectParses(ctx);
  await assertSnapshot(ctx, appState);
});

test("import foo/* as b", async (ctx) => {
  const { appState } = expectParses(ctx);
  await assertSnapshot(ctx, appState);
});

test(`import a/{ b, c/{d, e}, f/* }`, async (ctx) => {
  const { appState } = expectParses(ctx);
  await assertSnapshot(ctx, appState);
});

test("import ./foo/bar", async (ctx) => {
  const { appState } = expectParses(ctx);
  await assertSnapshot(ctx, appState);
});
