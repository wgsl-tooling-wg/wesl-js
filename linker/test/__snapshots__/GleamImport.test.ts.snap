export const snapshot = {};

snapshot[`import foo/bar 1`] = `
[
  {
    end: 14,
    imports: ImportTree {
      segments: [
        SimpleSegment {
          args: undefined,
          as: undefined,
          name: "foo",
        },
        SimpleSegment {
          args: undefined,
          as: undefined,
          name: "bar",
        },
      ],
    },
    kind: "treeImport",
    start: 0,
  },
]
`;

snapshot[`import foo/* as b 1`] = `
[
  {
    end: 17,
    imports: ImportTree {
      segments: [
        SimpleSegment {
          args: undefined,
          as: undefined,
          name: "foo",
        },
        Wildcard {
          as: "b",
        },
      ],
    },
    kind: "treeImport",
    start: 0,
  },
]
`;

snapshot[`import a/{ b, c/{d, e}, f/* } 1`] = `
[
  {
    end: 29,
    imports: ImportTree {
      segments: [
        SimpleSegment {
          args: undefined,
          as: undefined,
          name: "a",
        },
        SegmentList {
          list: [
            ImportTree {
              segments: [
                SimpleSegment {
                  args: undefined,
                  as: undefined,
                  name: "b",
                },
              ],
            },
            ImportTree {
              segments: [
                SimpleSegment {
                  args: undefined,
                  as: undefined,
                  name: "c",
                },
                SegmentList {
                  list: [
                    ImportTree {
                      segments: [
                        SimpleSegment {
                          args: undefined,
                          as: undefined,
                          name: "d",
                        },
                      ],
                    },
                    ImportTree {
                      segments: [
                        SimpleSegment {
                          args: undefined,
                          as: undefined,
                          name: "e",
                        },
                      ],
                    },
                  ],
                },
              ],
            },
            ImportTree {
              segments: [
                SimpleSegment {
                  args: undefined,
                  as: undefined,
                  name: "f",
                },
                Wildcard {
                  as: undefined,
                },
              ],
            },
          ],
        },
      ],
    },
    kind: "treeImport",
    start: 0,
  },
]
`;

snapshot[`import ./foo/bar 1`] = `
[
  {
    end: 16,
    imports: ImportTree {
      segments: [
        SimpleSegment {
          args: undefined,
          as: undefined,
          name: ".",
        },
        SimpleSegment {
          args: undefined,
          as: undefined,
          name: "foo",
        },
        SimpleSegment {
          args: undefined,
          as: undefined,
          name: "bar",
        },
      ],
    },
    kind: "treeImport",
    start: 0,
  },
]
`;
